### PR TITLE
Tdl 22160 add integration tests

### PR DESCRIPTION
**Pull request by kethan1122**
# Description of change
Adds following integration tests

- Discovery
- Pagination
- All fields
- Automatic fields
- Bookmarks
- Start_date

# Manual QA steps
 - Ran all the tests on dev-vm and made sure all of them pass
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

**Review by RushiT0122 — CHANGES_REQUESTED** [review]
Refer review comments on this [tap-iterable PR#22](https://github.com/singer-io/tap-iterable/pull/22)

File: tests/test_all_fields.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                for fields in all_record_fields:
+                    actual_all_fields.update(fields)
+
+                # Verify that more than just the automatic fields are replicated for each stream
+                self.assertTrue(expected_automatic_fields.issubset(actual_all_fields),
+                                msg=f'{expected_automatic_fields-actual_all_fields} is not in "expected_all_keys"')
+                self.assertTrue(
+                    expected_automatic_fields.issubset(actual_all_fields),
+                    msg=f'{expected_automatic_fields-actual_all_fields} is not in "expected_all_keys"',
+                )
```
Comment (RushiT0122) [inline tests/test_all_fields.py:85]: This looks like repeated assertion.

File: tests/test_automatic_fields.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                # expected values
+                expected_primary_keys = self.expected_primary_keys()
+                expected_keys = self.expected_automatic_fields().get(stream)
+
+                # Verify that you get some records for each stream
+                self.assertGreater(
+                    record_count_by_stream.get(stream, -1),
+                    0,
+                    msg="The number of records is not over the stream max limit",
+                )
```
Comment (RushiT0122) [inline tests/test_automatic_fields.py:56]:
```suggestion
                self.assertGreater(
                    record_count_by_stream.get(stream, -1), 0,
                    msg="The number of records is not over the stream max limit",
                )
```

File: tests/test_automatic_fields.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                records_pks_list = [
+                    tuple(message.get(pk) for pk in expected_primary_keys[stream])
+                    for message in [json.loads(t) for t in {json.dumps(d) for d in records}]
+                ]
+
+                # Remove duplicate primary keys
+                records_pks_set = set(records_pks_list)
+
+                # Verify there are no duplicate records
+                self.assertEqual(len(records), len(records_pks_set), msg=f"{stream} contains duplicate records")
```
Comment (RushiT0122) [inline tests/test_automatic_fields.py:87]: Just to be consistent with other assertion.
```suggestion
                self.assertEqual(
                    len(records), len(records_pks_set), 
                    msg=f"{stream} contains duplicate records")
```

File: tests/test_bookmarks.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+        second_sync_record_count = self.run_and_verify_sync(conn_id)
+        second_sync_records = runner.get_records_from_target_output()
+        second_sync_bookmarks = menagerie.get_state(conn_id)
+
+        ########################
+        # Test by Stream
+        ########################
+
+        for stream in streams_to_test:
```
Comment (RushiT0122) [inline tests/test_bookmarks.py:68]:
```suggestion
        ########################

        # Verify currently syncing is set to None after successful sync
        self.assertNotIn("currently_syncing", first_sync_bookmarks)
        self.assertNotIn("currently_syncing", second_sync_bookmarks)

        for stream in streams_to_test:
```

File: tests/test_bookmarks.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+                else:
+                    raise NotImplementedError(
+                        "INVALID EXPECTATIONS\t\tSTREAM: {} REPLICATION_METHOD: {}".format(
+                            stream, expected_replication_method
+                        )
+                    )
+
+                # Verify at least 1 record was replicated in the second sync
+                self.assertGreater(second_sync_count, 0, msg=f"We are not fully testing bookmarking for {stream}")
```
Comment (RushiT0122) [inline tests/test_bookmarks.py:159]: If this assertion doesn't pass then test will fail #82 with KeyError, so move this above.

File: tests/test_bookmarks.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        second_sync_bookmarks = menagerie.get_state(conn_id)
+
+        ########################
+        # Test by Stream
+        ########################
+
+        for stream in streams_to_test:
+            with self.subTest(stream=stream):
+
+                # Expected values
```
Comment (RushiT0122) [inline tests/test_bookmarks.py:71]:
```suggestion
                # verify that we get some records for each stream
                self.assertGreater(second_sync_record_count.get(stream), 0)

                # Expected values
```

File: tests/test_bookmarks.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+                    )
+
+                elif expected_replication_method == self.FULL_TABLE:
+
+                    # Verify the syncs do not set a bookmark for full table streams
+                    self.assertIsNone(first_bookmark_value)
+                    self.assertIsNone(second_bookmark_value)
+
+                    # Verify the number of records in the second sync is the same as the first
+                    self.assertEqual(second_sync_count, first_sync_count)
```
Comment (RushiT0122) [inline tests/test_bookmarks.py:149]:
```suggestion
                    self.assertEqual(second_sync_count, first_sync_count)

                    # Verify both syncs have same records
                    for record in first_sync_messages:
                        self.assertIn(record, second_sync_messages)
```

File: tests/test_pagination.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+                records_pks_set = {
+                    tuple(message.get("data").get(primary_key) for primary_key in primary_keys)
+                    for message in sync_messages
+                }
+                records_pks_list = [
+                    tuple(message.get("data").get(primary_key) for primary_key in primary_keys)
+                    for message in sync_messages
+                ]
+
+                self.assertCountEqual(records_pks_set, records_pks_list, msg=f"We have duplicate records for {stream}")
```
Comment (RushiT0122) [inline tests/test_pagination.py:56]:
```suggestion
                records_pks_list = [
                    tuple(message.get("data").get(primary_key) for primary_key in primary_keys)
                    for message in sync_messages
                ]

                self.assertCountEqual(set(records_pks_list), records_pks_list, msg=f"We have duplicate records for {stream}")
```

Comment (RushiT0122) [reply tests/test_pagination.py:56]: Along with this we also need to test uniqueness of records between all the pages.

**Review by RushiT0122 — CHANGES_REQUESTED** [review]
Added few suggestions.

File: tests/test_all_fields.py (hunk 1 above)
Comment (kethan1122) [reply tests/test_all_fields.py]: Removed

File: tests/test_automatic_fields.py (hunk 2 above)
Comment (kethan1122) [reply tests/test_automatic_fields.py]: changed

File: tests/test_automatic_fields.py (hunk 3 above)
Comment (kethan1122) [reply tests/test_automatic_fields.py]: Changed

File: tests/test_bookmarks.py (hunk 4 above)
Comment (kethan1122) [reply tests/test_bookmarks.py]: Added

File: tests/test_bookmarks.py (hunk 5 above)
Comment (kethan1122) [reply tests/test_bookmarks.py]: Moved it

File: tests/test_bookmarks.py (hunk 6 above)
Comment (kethan1122) [reply tests/test_bookmarks.py:75]: We have this assertion just below

File: tests/test_bookmarks.py (hunk 7 above)
Comment (kethan1122) [reply tests/test_bookmarks.py:157]: added

**Comment by kethan1122** [comment]
responded to all comments

**Review by RushiT0122 — APPROVED** [review]

**Pull request merged**